tablo/interfaces/arcgis/views.py

```diff
@@ -259,44 +266,61 @@ def handle_request(self, request, **kwargs):
         except (DatabaseError, ValueError):
             return HttpResponseBadRequest(json.dumps({'error': 'Invalid request'}))
 
+        # Process limited query before calculating totals and counts
+        exceeded_limit = len(query_response) > int(limit)
+        query_response = query_response[:-1] if exceeded_limit else query_response
+
         if return_count_only:
-            response = query_response[0]
+            data = query_response[0]
+        elif return_format == 'csv':
+            header = list(query_response[0].keys())
+            header.sort()
+
+            data = [[h.strip('"').join('""') for h in header]] if offset == 0 else []
+            for item in query_response:
+                data.append([str(item[field]).strip('"').join('""') for field in header])
         else:
-            record_count = len(query_response)
-            exceeded_limit = record_count > int(limit)
-            query_response = query_response[:-1] if exceeded_limit else query_response
-
-            response = {
+            data = {
                 'count': len(query_response),
                 'total': len(query_response),
                 'exceededTransferLimit': exceeded_limit
             }
             if return_ids_only:
                 object_id_field = self.feature_service_layer.object_id_field
-                response['objectIdFieldName'] = object_id_field
-                response['objectIds'] = [feature[object_id_field] for feature in query_response]
+                data['objectIdFieldName'] = object_id_field
+                data['objectIds'] = [feature[object_id_field] for feature in query_response]
             else:
                 features = convert_wkt_to_esri_feature(query_response, self.feature_service_layer)
-                response.update({
+                data.update({
                     'count': len(features),        # Root level (source) features
                     'total': len(query_response),  # Total number of records queried
                     'fields': self.feature_service_layer.fields,
                     'geometryType': 'esriGeometryPoint',
                     'features': features
                 })
 
-                if response['total'] > response['count']:
-                    response['relatedFields'] = {
+                if data['total'] > data['count']:
+                    data['relatedFields'] = {
                         r.related_title: r.fields for r in self.feature_service_layer.relations
                     }
 
-        content = json.dumps(response, default=json_date_serializer)
-        content_type = 'application/json'
-        if self.callback:
-            content = '{callback}({data})'.format(callback=self.callback, data=content)
-            content_type = 'text/javascript'
+        if return_format == 'csv':
+            content = '\n'.join(','.join(col if isinstance(col, str) else str(col) for col in row) for row in data)
```

Comment:
I'm not worried about the extra import, it is in the std lib.  The overhead of that and the `with` block may be minor compared to performance improvements we get from using the lib - most of it is implemented in [C](http://svn.python.org/projects/python/trunk/Modules/_csv.c).

It may be worth doing a small comparison to see how performance differs between your version and the `csv` module version for a large-ish number of rows.